pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SingleSnapshotAbortedTxnProcessorImpl.java

```diff
@@ -54,10 +56,20 @@ public SingleSnapshotAbortedTxnProcessorImpl(PersistentTopic topic) {
                 .getTransactionBufferSnapshotServiceFactory()
                 .getTxnBufferSnapshotService().getReferenceWriter(TopicName.get(topic.getName()).getNamespaceObject());
         this.takeSnapshotWriter.getFuture().exceptionally((ex) -> {
-                    log.error("{} Failed to create snapshot writer", topic.getName());
-                    topic.close();
-                    return null;
-                });
+            log.error("{} Failed to create snapshot writer", topic.getName());
+            // Don't directly use the topic object to close, because the topicFuture might not
+            // be completed at that time, which could leave closed topics in the cache(at BrokerService).
+            CompletableFuture<Optional<Topic>> topicFuture = topic.getBrokerService().getTopics().get(topic.getName());
+            if (topicFuture != null) {
+                topicFuture.thenAccept(t -> t.ifPresent(v -> {
```

Comment:
I'm not sure why the value of `topics` is an `Optional`. I just have a question that when could the value be `Optional.empty()`? If so, an empty `Optional` will be cached and never removed.